File: Data Structures & Algorithms/combination-target-sum/submission-2.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res=[]
        def backtrack(i,subset):
            if sum(subset)> target or i>=len(nums):
                return 
            if sum(subset)== target:
                res.append(subset.copy())
                return
            subset.append(nums[i])
            backtrack(i,subset)
            subset.pop()
            backtrack(i+1,subset)
            
        backtrack(0,[])
        return res

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-2.cpp
class Solution {
public:
    bool backtracking(vector<int>& nums,int i , int sum,int target){
        if(sum==target)
            return true;
        if (i>=nums.size())
            return false;
        return backtracking(nums,i+1,sum+nums[i],target)or backtracking(nums,i+1,sum,target);
    }
    bool canPartition(vector<int>& nums) {
        int sum = 0;
        for (int num : nums) {
            sum += num;
        }
        if (sum%2!=0)
            return false;
        
        return  backtracking(nums,0,0,sum/2);   
        }
};
